### PR TITLE
Add IPv6 Extension Fragment Header

### PR DESCRIPTION
**Pull request by hjr3**

### CODE REVIEW

File: src/wire/ipv6fragment.rs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    /// Return the length, in bytes, of a header that will be emitted from this high-level
+    /// representation.
+    pub fn buffer_len(&self) -> usize {
+        field::IDENT.end
+    }
+
+    /// Emit a high-level representation into an IPv6 Fragment Header.
+    pub fn emit<T: AsRef<[u8]> + AsMut<[u8]> + ?Sized>(&self, header: &mut Header<&mut T>) {
+        header.set_next_header(self.next_header);
+        header.set_reserved(0);
```
Comment (hjr3) [inline src/wire/ipv6fragment.rs]: I did not see an example of how to handle reserved fields. This is what I came up with.

File: src/wire/ipv6fragment.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    /// Set the next header field.
+    #[inline]
+    pub fn set_next_header(&mut self, value: Protocol) {
+        let data = self.buffer.as_mut();
+        data[field::NXT_HDR] = value.into();
+    }
+
+    /// Set the reserved field.
+    #[inline]
+    pub fn set_reserved(&mut self, value: u8) {
```
Comment (whitequark) [inline src/wire/ipv6fragment.rs]: Let's have a dedicated function that clears all reserved fields to their default values instead.

**Comment by whitequark** [comment]
Why do we need this header? smoltcp does not support fragmentation.

**Comment by hjr3** [comment]
@whitequark this is part of the IPv6 implementation issue https://github.com/m-labs/smoltcp/issues/103#issuecomment-357706855

Per [RFC 8200 Section 4](https://tools.ietf.org/html/rfc8200#section-4):

> A full implementation of IPv6 includes implementation of the
   following extension headers:
      Hop-by-Hop Options
      Fragment
      Destination Options
      Routing
      Authentication
      Encapsulating Security Payload

In talking with @dlrobertson, I believe the plan is to simply recognize these headers and then return the appropriate ICMPv6 response.

**Comment by whitequark** [comment]
OK

File: src/wire/ipv6fragment.rs (hunk 2 above)
Comment (hjr3) [reply src/wire/ipv6fragment.rs]: I made a single `set_reserved` function that sets both reserved fields to `0`.

File: src/wire/ipv6fragment.rs (hunk 1 above)
Comment (hjr3) [reply src/wire/ipv6fragment.rs]: resolved in https://github.com/m-labs/smoltcp/pull/168#discussion_r168937617

File: src/wire/ipv6fragment.rs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+// Format of the Fragment Header
+//
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+// |  Next Header  |   Reserved    |      Fragment Offset    |Res|M|
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+// |                         Identification                        |
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//
+// See https://tools.ietf.org/html/rfc8200#section-4.4 for details.
```
Comment (dlrobertson) [inline src/wire/ipv6fragment.rs]: nit: section 4.5 https://tools.ietf.org/html/rfc8200#section-4.5

File: src/wire/ipv6fragment.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    /// Set 8-bit reserved field after the next header field.
+    /// Set 2-bit reserved field between fragment offset and more fragments.
+    #[inline]
+    pub fn set_reserved(&mut self) {
+        let data = self.buffer.as_mut();
+
+        data[field::RESERVED] = 0;
+
+        // Retain the higher order 5 bits and lower order 1 bit
+        let raw = (0 & 0x6) | (data[3] & 0xf9);
```
Comment (dlrobertson) [inline src/wire/ipv6fragment.rs]: `let raw = data[3] & 0xf9` should be sufficient.

File: src/wire/ipv6fragment.rs [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        data[3] = raw;
+    }
+
+    /// Set the fragment offset field.
+    #[inline]
+    pub fn set_frag_offset(&mut self, value: u16) {
+        let data = self.buffer.as_mut();
+        // Retain the lower order 3 bits
+        let raw = ((value & 0x1fff) << 3) | ((data[3] & 0x3) as u16);
```
Comment (dlrobertson) [inline src/wire/ipv6fragment.rs]: `data[3] & 0x7` if you want to maintain the reserved bits. If you don't want to maintain the reserved bits `0x1` should be fine right?

Comment (hjr3) [reply src/wire/ipv6fragment.rs]: Yup, too many `3`'s in this line of code. Fixed.

File: src/wire/ipv6fragment.rs (hunk 4 above)
Comment (hjr3) [reply src/wire/ipv6fragment.rs]: Sure, fixed.

File: src/wire/ipv6fragment.rs (hunk 3 above)
Comment (hjr3) [reply src/wire/ipv6fragment.rs]: Oops, fixed.

File: src/wire/ipv6fragment.rs [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        Ok(header)
+    }
+
+    /// Ensure that no accessor method will panic if called.
+    /// Returns `Err(Error::Truncated)` if the buffer is too short.
+    pub fn check_len(&self) -> Result<()> {
+        let data = self.buffer.as_ref();
+        let len = data.len();
+
+        if len < field::IDENT.end {
```
[the remark this answers is not in the document]
Comment (hjr3) [reply src/wire/ipv6fragment.rs]: fixed

**Comment by whitequark** [comment]
@m-labs-homu r+

**Comment by m-labs-homu** [comment]
:pushpin: Commit 1746ddc has been approved by `whitequark`

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit 1746ddc58909c543ff514d0da0bd52b0c908a991 with merge dbaabd7304087301680a5a9ed312f9235e71deda...

**Comment by m-labs-homu** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/m-labs/smoltcp/builds/344669020?utm_source=github_status&utm_medium=notification)

**Comment by whitequark** [comment]
@m-labs-homu retry

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit 1746ddc58909c543ff514d0da0bd52b0c908a991 with merge e114c46fce2283ef452c78c8bba60346e0dcbd29...

**Pull request closed without merging**